<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Estimate the cause of recurrent vivax malaria using genetic data • Pv3Rs</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Estimate the cause of recurrent vivax malaria using genetic data">
<meta name="description" content="Estimate per-person posterior probabilities that each recurrent Plasmodium vivax (Pv) infection is a recrudescence, relapse, reinfection (3Rs) using per-person P. vivax genetic data on two or more infections.">
<meta property="og:description" content="Estimate per-person posterior probabilities that each recurrent Plasmodium vivax (Pv) infection is a recrudescence, relapse, reinfection (3Rs) using per-person P. vivax genetic data on two or more infections.">
<meta property="og:image" content="https://aimeertaylor.github.io/Pv3Rs/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">Pv3Rs</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/demonstrate-usage.html">Demonstrate Pv3Rs usage</a></li>
    <li><a class="dropdown-item" href="articles/enumerate.html">Enumerate graphs and partitions</a></li>
    <li><a class="dropdown-item" href="articles/graph-prior.html">Understand graph prior ramifications</a></li>
    <li><a class="dropdown-item" href="articles/half-siblings.html">Understand half-sibling misspecification</a></li>
    <li><a class="dropdown-item" href="articles/intra-episode-siblings.html">Explore treatment of intra-episode siblings</a></li>
    <li><a class="dropdown-item" href="articles/understand-posterior.html">Understand posterior estimates</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/aimeertaylor/Pv3Rs/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="pv3rs-">Pv3Rs <a class="anchor" aria-label="anchor" href="#pv3rs-"></a>
</h1>
</div>
<p>An R package developed to enable statistical genetic inference of <em>Plasmodium vivax</em></p>
<ul>
<li>Relapse</li>
<li>Reinfection</li>
<li>Recrudescence</li>
</ul>
<p>The main function <code><a href="reference/compute_posterior.html">Pv3Rs::compute_posterior</a></code> can be used to estimate posterior probabibilities of relapse, reinfection and recrudescence for one or more recurrence using genetic data to update a prior, which is ideally informative (e.g., based on time-to-event information).</p>
<p>Other features are more general:</p>
<ul>
<li><p><code><a href="reference/plot_data.html">Pv3Rs::plot_data</a></code> can be used to visualise genetic data for molecular correction regardless of the analytical method (e.g., it could be used to visualise <em>Plasmodium falciparum</em> data intended for analysis using a WHO match counting algorithm).</p></li>
<li><p>Together, <code><a href="reference/plot_simplex.html">Pv3Rs::plot_simplex</a></code> and <code><a href="reference/project2D.html">Pv3Rs::project2D</a></code> can be used to plot a vector of any three numbers between 0 and 1 that sum to one.</p></li>
</ul>
<div class="section level2">
<h2 id="please-be-aware-of-the-following-points">Please be aware of the following points!<a class="anchor" aria-label="anchor" href="#please-be-aware-of-the-following-points"></a>
</h2>
<div class="section level3">
<h3 id="unstable-state">Unstable state:<a class="anchor" aria-label="anchor" href="#unstable-state"></a>
</h3>
<p>The R package is currently under development and thus liable to contain many errors. The model around which the Pv3Rs R package is being developed is documented in the preprint [1]. The model builds upon a prototype documented in [2] (see [1] to better understand how the model underpinning Pv3Rs differs from the prototype).</p>
<p>[1] <a href="https://www.medrxiv.org/content/10.1101/2022.11.23.22282669v1" class="external-link">Taylor, Foo &amp; White, 2022</a></p>
<p>[2] <a href="https://www.nature.com/articles/s41467-019-13412-x" class="external-link">Taylor &amp; Watson et al. 2019</a></p>
</div>
<div class="section level3">
<h3 id="prior-considerations">Prior considerations:<a class="anchor" aria-label="anchor" href="#prior-considerations"></a>
</h3>
<ul>
<li><p>Genetic data are modelled using a Bayesian model, whose prior is ideally informative (in [2] prior estimates were generated by a time-to-event model built by James Watson) because the cause of recurrent <em>P. vivax</em> malaria is not always identifiable from genetic data alone (when the data are consistent with recurrent parasites that are relatively unrelated to those in all preceding infections, both reinfection and relapse are plausible; meanwhile, when the data are compatible with recurrent parasites that are clones of those in the preceding infection, both recrudescence and relapse are plausible).</p></li>
<li><p>The main Pv3Rs function, <code><a href="reference/compute_posterior.html">compute_posterior()</a></code>, could be used to estimate the probable cause of recurrent <em>P. falciparum</em> malaria by setting the prior probability of relapse to zero. However, the current version of Pv3Rs is suboptimal for <em>P. falciparum</em> recurrent state inference: genotyping errors, which are not accounted for under the current Pv3Rs model, are liable to lead to the misclassification of recrudescence as reinfection when the probability of relapse is zero <em>a priori</em>.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="notable-assumptions">Notable assumptions:<a class="anchor" aria-label="anchor" href="#notable-assumptions"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Relationship graphs compatible with a given recurrent state sequence are equally likely <em>a priori</em>
</li>
<li>Perfect detection of alleles (no genotyping error)</li>
<li>Perfect detection of parasites (problematic for low density clones)</li>
<li>Perfect detection of episodes (requires active follow up)</li>
<li>Mutually exclusive recurrent states (requires frequent follow up with treatment)</li>
<li>No within-host <em>de novo</em> mutations</li>
<li>Parasites are outbred (implies the parasite population is infinitely large and panmitic)</li>
<li>Genetic markers are independent
<ul>
<li>Ignores long-range linkage disequilibrium due to background relatedness</li>
<li>Ignores short-range linkage disequilbrium between between sibling parasites</li>
</ul>
</li>
<li>All siblings are regular siblings
<ul>
<li>Siblings are transitive (not true of some parent child-like sibling trios)</li>
<li>Siblings are independent (not true of meiotic siblings)</li>
<li>Siblings draw from at most two parents (not true of half-siblings)</li>
</ul>
</li>
</ol>
<p>The first assumption listed above has a small artefactual effect on posterior estimates when relationship graphs grow in size. This is demonstrated in one of the examples of <code><a href="reference/compute_posterior.html">Pv3Rs::compute_posterior()</a></code> and will be explained in more detail in an upcoming vignette.</p>
<p>For studies with possibly high rates of recrudescence, we recommend a sensitivity analysis to explore the impact of genotyping errors on recrudescence. An example will be provided in an upcoming vignette.</p>
<p>The marker independence assumption is liable to lead to overconfident posterior probabilities when markers are linked.</p>
</div>
<div class="section level3">
<h3 id="computational-limits">Computational limits:<a class="anchor" aria-label="anchor" href="#computational-limits"></a>
</h3>
<ul>
<li><p>Pv3Rs scales to 100s of markers but not whole-genome sequence (WGS) data.</p></li>
<li><p>We do not recommend running <code><a href="reference/compute_posterior.html">Pv3Rs::compute_posterior()</a></code> for data whose total genotype count (sum of per-episode multiplicities of infection) exceeds eight. If the total genotype counts exceeds eight but there are multiple recurrences, it might be possible to generate recurrent state estimates for recurrences analysing episodes pairwise (this approach was used in [2]).</p></li>
<li><p>We have not tested the per-marker allele limit of <code><a href="reference/compute_posterior.html">Pv3Rs::compute_posterior()</a></code>. Very high marker cardinalities could lead to very small allele frequencies and thus underflow problems.</p></li>
</ul>
</div>
<div class="section level3">
<h3 id="population-level-allele-frequencies">Population-level allele frequencies<a class="anchor" aria-label="anchor" href="#population-level-allele-frequencies"></a>
</h3>
<p><code><a href="reference/compute_posterior.html">compute_posterior()</a></code> requires population-level alelle frequencies. To avoid bias due to within-host selection of recrudescent parasites, we recommended using only enrolment episodes to estimate population-level allele frequencies. That said, if most recurrences are either reinfections or relapses, both of which are draws from the mosquito population (albeit a delayed draw in the case of a relapse), in the absence of systematic within-patient selection (as might occur when break-through infections encounter lingering drug pressure), estimates based on all episodes should be unbiased and more precise than those based on enrolment episodes only.</p>
</div>
<div class="section level3">
<h3 id="other-points">Other points:<a class="anchor" aria-label="anchor" href="#other-points"></a>
</h3>
<p>Unfortunately, the Pv3Rs model does not exploit read count data at present. However, read count data could be used to compute population-level allele frequencies, assuming they are not biased by experimental artefacts.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install or update latest stable version of devtools from CRAN</span></span>
<span><span class="co"># I highly recommend doing this if you've recently updated R and RStudio to versions 4.3.0 and 2023.3.1.446, respectively; </span></span>
<span><span class="co"># otherwise, you might encounter problems rendering documentation</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install R.rsp, a vignette builder used to build LaTeX vignettes in Pv3Rs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"R.rsp"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install Pv3Rs from GitHub </span></span>
<span><span class="co"># I highly recommend doing this in RStudio as RStudio installs pandoc needed to build vignettes</span></span>
<span><span class="co"># If you're working in R outside of RStudio you might need to install pandoc and check its path; </span></span>
<span><span class="co"># otherwise set build_vignettes = FALSE</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"aimeertaylor/Pv3Rs"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load and attach the package</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/aimeertaylor/Pv3Rs" class="external-link">Pv3Rs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># View documentation and examples for main function</span></span>
<span><span class="op">?</span><span class="va">compute_posterior</span></span>
<span></span>
<span><span class="co"># Load the demo vignette</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span><span class="st">"demo"</span>, package <span class="op">=</span> <span class="st">"Pv3Rs"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Lists available functions, as well as example data sets and their documentation [check]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/help.html" class="external-link">help</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"Pv3Rs"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/aimeertaylor/Pv3Rs/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/aimeertaylor/Pv3Rs/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Pv3Rs</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Aimee Taylor <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2337-8992" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a>  </li>
<li>Yong See Foo <br><small class="roles"> Author </small>   </li>
<li>European Union, Project 101110393 — PvRecur <br><small class="roles"> Funder </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aimee Taylor, Yong See Foo, European Union, Project 101110393 — PvRecur.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
