# Pv3Rs <img src="man/figures/logo.png" align="right" alt="" width="120" />

An R package for *Plasmodium vivax* molecular correction: statistical genetic
inference of *P. vivax*

- Relapse
- Reinfection
- Recrudescence

The core function `compute_posterior()` can be used to estimate per-person
posterior probabilities of relapse, reinfection and recrudescence using *P.
vivax* genetic data on paired samples from one or more recurrence to update a
prior, which is ideally informative (e.g., based on time-to-event information).

Two other important features are more general:

- `plot_data()` can be used to visualise genetic data for molecular
correction regardless of the analytical method; for example, *Plasmodium
falciparum* data intended for analysis using a WHO match counting algorithm.

- `plot_simplex()` can be used to visualise per-recurrence estimates of the
probabilities of relapse, reinfection and recrudescence or a vector of any three
numbers in zero to one that sum to one.

## Please be aware of the following points!

### Pre-release state: 

The R package is still under development and may contain some errors. Moreover,
the statistical model around which the Pv3Rs R package is being developed is not
yet peer-reviewed and thus liable to modification. The model is documented in
the preprint [1]. The model builds upon a prototype documented in [2] (see [1]
to better understand how the model underpinning Pv3Rs differs from the
prototype).

[1] [Taylor, Foo & White, 2022](https://www.medrxiv.org/content/10.1101/2022.11.23.22282669v1)

[2] [Taylor & Watson et al. 2019](https://www.nature.com/articles/s41467-019-13412-x)

### Prior considerations: 

- Genetic data are modelled using a Bayesian model, whose prior is ideally
informative (in [2] prior estimates were generated by a time-to-event model
built by James Watson) because the cause of recurrent *P. vivax* malaria is not
always identifiable from genetic data alone: when the data are consistent with
recurrent parasites that are relatively unrelated to those in all preceding
infections, both reinfection and relapse are plausible; meanwhile, when the data
are compatible with recurrent parasites that are clones of those in the
preceding infection, both recrudescence and relapse are plausible.

- The main Pv3Rs function, `compute_posterior()`, could be used to estimate the
probable cause of recurrent *P. falciparum* malaria by setting the prior
probability of relapse to zero. However, the current version of Pv3Rs is
suboptimal for *P. falciparum* molecular correction: genotyping errors, which
are not accounted for under the current Pv3Rs model, are liable to lead to the
misclassification of recrudescence as reinfection when the probability of
relapse is zero *a priori* (and of recrudescence as relapse when the prior
probability of relapse exceeds zero). 

### Notable limitations: 

Limitation | Reason
----------- | ------
Possible misclassification of relapse as reinfection | Not modelling errors and half-sibling misspecification
Possible misclassification of recrudescence as relapse | Not modelling errors
Possible inconsistency with data on more-and-more markers | Not modelling errors
Possible misclassification of persistent and/or accumulated events (e.g., recrudescence plus reinfection) as relapse | Modelling 3Rs as mutually exclusive events
Strong prior impact of probabilities of reinfection / recrudescence | Assumption that relationship graphs compatible with a given recurrent state sequence are equally likely *a priori*

<!-- Because, we don't model population structure, we also recommend a
3) Perfect detection of parasites (problematic for low density clones)
7) Parasites are outbred (implies the parasite population is infinitely large and panmitic)
For studies with possibly high rates of recrudescence, we recommend a
sensitivity analysis to explore the impact of genotyping errors on
recrudescence. An example will be provided in an upcoming manuscript. -->

### Computational limits:

- Pv3Rs scales to 100s of markers but not whole-genome sequence (WGS) data.  

- We do not recommend running `compute_posterior()` for data whose total 
genotype count (sum of per-episode multiplicities of infection) exceeds eight. 
If the total genotype counts exceeds eight but there are multiple recurrences,
it might be possible to generate recurrent state estimates for recurrences
analysing episodes pairwise (this approach was used in [2]).

- We have not tested the per-marker allele limit of `compute_posterior()`. 
Very high marker cardinalities could lead to very small allele frequencies and 
thus underflow problems. 


### Population-level allele frequencies: 

In addition to paired data, `compute_posterior()` requires as input
population-level allele frequencies. To avoid bias due to within-host selection
of recrudescent parasites, we recommend using only enrolment episodes to
estimate population-level allele frequencies. That said, if most recurrences are
either reinfections or relapses, both of which are draws from the mosquito
population (albeit a delayed draw in the case of a relapse), assuming there is
no systematic within-patient selection (as might occur when infections encounter
lingering drug pressure), estimates based on all episodes should be unbiased and
more precise than those based on enrolment episodes only.

### Read-count data: 

Unfortunately, the Pv3Rs model does not exploit read count data at present.
However, read count data could be used to compute population-level allele
frequencies, assuming they are not biased by experimental artefacts.


## Installation 

```r
# Install or update latest stable version of devtools from CRAN
install.packages("devtools")

# Install R.rsp, a vignette builder used to build LaTeX vignettes in Pv3Rs
install.packages("R.rsp")

# Install Pv3Rs from GitHub 
# I highly recommend doing this in RStudio as RStudio installs pandoc needed to build vignettes.
# If you're working in R outside of RStudio you might need to install pandoc and check its path; 
# otherwise set build_vignettes = FALSE
devtools::install_github("aimeertaylor/Pv3Rs", build_vignettes = TRUE)

# Load and attach Pv3Rs
library(Pv3Rs)

# List links to all available documentation
help(package = "Pv3Rs")

# List links to vignettes
vignette(package = "Pv3Rs")

# View function documentation, e.g., 
?compute_posterior
```
