---
title: "Demonstrating Pv3Rs usage"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Demonstrating Pv3Rs usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align ='center', 
  fig.width = 7
)
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup}
library(Pv3Rs)
```

## Synthetic example

We first look at a synthetic example of three episodes (names are optional) with three markers (names are obligatory) whose alleles have known frequencies. 

```{r}
y <- list("Enrollment" = list(m1 = c('C','G','T'), 
                              m2 = c('A','C'),
                              m3 = c('C','G','T')),
          "Recurrence 1" = list(m1 = c('C','T'), 
                                m2 = c('A'), 
                                m3 = c('A','C')),
          "Recurrence 2" = list(m1 = c('T'), 
                                m2 = c('A'), 
                                m3 = c('A')))

fs <- list(m1 = c(A = 0.27, C = 0.35, G = 0.18, T = 0.20),
           m2 = c(A = 0.78, C = 0.14, G = 0.07, T = 0.01),
           m3 = c(A = 0.21, C = 0.45, G = 0.26, T = 0.08))
```

We can plot the data using `plot_data()`. 

```{r, fig.align='center'}
plot_data(ys = list("Data" = y), fs = fs)
```
Different colours in the data plot (left) represent different alleles. Alleles
are also listed in the legend (right), with one column per marker. The marker
order in the legend is the same as in the data plot (see horizontal axis). When
allele frequencies are specified, common alleles have relatively large legend
areas, while rare alleles have relatively small legend areas.

Given the number of alleles observed during each episode, the most parsimonious
explanation is that there are three, two, and one distinct parasite genotypes
in the three episodes respectively (a genotype is defined here as a
realisation of the haploid parasite genome). 

```{r}
determine_MOIs(y)
```

If better MOI estimates are available, they can be used in genetic recurrent
state inference by specifying the `MOI` argument in the function
`compute_posterior()`.

When performing genetic recurrent state inference, the bulk of the computational
time lies in computing log-likelihoods of graphs of relationships between
genotypes.

```{r, cache=TRUE}
post <- compute_posterior(y, fs, return.RG = TRUE, return.logp = TRUE)
```

In the call to `compute_posterior()` above we did not specify a prior, and so a
uniform prior across all three recurrence states per recurrence is assumed by
default. We set `return.RG=TRUE` for relationship graphs to be returned. This is
by default `FALSE` to conserve memory.

The marginal posterior probabilities of each recurrence state is stored in
`post$marg`.

```{r}
post$marg
```

The marginal posterior probabilities can be plotted on the simplex using the
function `plot_simplex()`.

```{r}
# Plot simplex
par(mar = rep(0.1,4))
plot_simplex(v_labels = c(C = "Recrudescence", 
                          L = "Relapse", 
                          I = "Reinfection")[colnames(post$marg)], 
             v_cutoff = 0.75)

# Project marginal posterior probabilties onto 2D coordinates
xy <- apply(post$marg, 1, project2D) 
xy_prior <- project2D(rep(1/3,3)) 

# Plot projection on the simplex
points(x = xy["x",], y = xy["y",], pch = 20) 
text(x = xy["x",], y = xy["y",], pos = 3, labels = gsub("_", " ", rownames(post$marg)))

# Add default prior per recurrence
points(x = xy_prior["x"], y = xy_prior["y"], pch = 20)
text(x = xy_prior["x"], y = xy_prior["y"], pos = 3, labels = "Prior probability")
```

The point in the yellow region is most likely a recrudescence with posterior
probability greater than 0.5 (it falls in the bright yellow region); the point
in the red region is most likely a reinfection with posterior probability
greater than 0.5 (it falls in the bright red region).

We can access the joint posterior through `post$joint`. Here, we find the most
likely sequence of recurrence states:

```{r}
post$joint[which.max(post$joint)]
```

The log-likelihood of each relationship graph is also returned in the output.
We plot the relationship graph with the largest likelihood. 

```{r}
# Extract the relationship graph (RG) with the largest log likelihood
RG <- post$RGs[[which.max(sapply(post$RGs, function(RG) RG$logp))]]

# Generate genotype names for plot_RG()
gs <- paste0("g", 1:6) 

# Generate episode indices for plot RG()
ts_per_gs <- rep(1:length(y), determine_MOIs(y)) 

# Plot RG
par(mar = rep(0.1,4))
plot_RG(RG_to_igraph(RG, gs, ts_per_gs), edge.curved=0.25, vertex.size=20)

# Generate infection colours in order to add a legend
infection_colours <- RColorBrewer::brewer.pal(n = 8, "Set2") 

# Add a legend
legend("left", pch = 21, pt.bg = infection_colours[unique(ts_per_gs)], 
       bty = "n", legend = names(y), title = "Episode")

# The sequences of states compatible with the relationship graph RG
seqs_comp_MLE_RG <- compatible_rstrs(RG, split(gs, ts_per_gs))
```

N.B. This approach of finding the relationship graph for which the data are most
probable does not take into account symmetries in the graphs. Specifically,
relationship graphs that are isomorphic up to within-episode genotype
permutations have the same likelihood. The user should note that the
relationship graph with the largest likelihood might not be in the equivalence
class for which the data are most probable when all relationship graphs in that
class are summed over.

The figure below shows histograms of the likelihoods of the relationship graphs
compatible with each possible sequence of recurrence states, along with the
likelihood of the sequence of recurrence states (red vertical line). Although
sequences `r seqs_comp_MLE_RG` are all associated with the relationship graph
with the largest log-likelihood, we observe that the likelihoods of these
sequences of recurrence states decrease as the number of compatible relationship
graphs increase. This behaviour is expected due to our choice of a uniform prior
over relationship graphs.

```{r, fig.height=10}
# Extract sequences
seqs <- names(post$joint)

# Extract likelihoods
prob_data_given_graph <- exp(sapply(post$RGs, function(RG) RG$logp))

# For each graph, extract compatible sequences and store in a list
gs_per_ts <- split(gs, ts_per_gs)
seq_per_RG_list <- lapply(post$RGs, compatible_rstrs, gs_per_ts)

# For each sequence, return TRUE if graph ID compatible 
seq_per_RG_matx <- sapply(seqs, function(seq) sapply(seq_per_RG_list, function(x) seq %in% x))

# For each sequence with non-zero posterior probability, plot histogram of likelihoods
poss_seqs <- names(sort(post$joint[post$joint > 0], decreasing = T))
likelihoods <- array(0, dim = c(1, length(post$joint)), dimnames = list(NULL, names(post$joint)))

par(mfrow = c(3, 2))
ML <- max(prob_data_given_graph)
xlims <- range(prob_data_given_graph)
for(poss_seq in poss_seqs) {
  hist(prob_data_given_graph[seq_per_RG_matx[,poss_seq]], breaks = 100, 
       xlim = xlims, xlab = "Likelihood", col = "cornflowerblue", border = "cornflowerblue", 
       main = sprintf("%s (compatible with %s relationship graphs)", poss_seq, 
                      sum(seq_per_RG_matx[,poss_seq])))
  likelihoods[1,poss_seq] <- mean(prob_data_given_graph[seq_per_RG_matx[,poss_seq]])
  abline(v = likelihoods[1,poss_seq], col = "red")
}
```

Note that because the prior on sequences is uniform in the current case, the likelihood is proportional to the posterior, and so the maximum posterior estimate is equivalent to the maximum likelihood estimate (see scatter plot below). 

```{r, fig.height=5}
par(mfrow = c(1, 1))
plot(y = post$joint, x = likelihoods[1,], bty = "n",
     ylab = "Posterior", xlab = "Likelihood", pch = 20)
lmfit <- lm(post$joint ~ likelihoods[1,])
abline(a = 0, b = lmfit$coefficients[2], col = "blue")
text(y = post$joint[post$joint > 0], x = likelihoods[1,post$joint > 0], pos = 2, 
     labels = names(post$joint[post$joint >0]))
```

## Microsatellite data

Next, we apply our methods to real *P. vivax* microsatellite data available [here](https://github.com/jwatowatson/RecurrentVivax/blob/master/RData/GeneticModel/MS_data_PooledAnalysis.RData).

```{r}
load('../data/MS_data_PooledAnalysis.RData')
MSs_all <- tail(names(MS_pooled), 8) # marker names
head(MS_pooled, 10) # first 10 rows
```

In this example, we use a Empirical Bayes estimate of allele frequencies inferred from the alleles observed in the enrolment episodes.

```{r}
D_weight_Prior <- 1 # uniform Dirichlet prior for allele frequencies
ind_enrol = which(MS_pooled$Episode==1) # Indices of enrolment episodes
Alpha_Posteriors <- apply(MS_pooled[,MSs_all], 2, function(x, ind_enrol){
  xmax <-  max(x,na.rm=T) # number of possible alleles for this marker
  # prior concentration parameters
  param_vector <- array(D_weight_Prior, dim=xmax, dimnames=list(1:xmax))
  # observed data summarised as counts
  obs_counts <- table(x[ind_enrol])
  # posterior concentration parameters
  param_vector[names(obs_counts)] <-  param_vector[names(obs_counts)] + obs_counts
  return(param_vector)
}, ind_enrol)

# Posterior mean 
Fs_Combined <- sapply(Alpha_Posteriors, function(x){x/sum(x)})
```

We group rows of the full dataset by patient ID, and filter out cases that are not currently suitable for inference.

```{r, results="hold"}
library(plyr)
y.dfs <- plyr::dlply(MS_pooled, 'ID')
no_recur <- c() # Discard patient if no recurrent data
too_many_genos <- c() # Discard patient if more than eight genotypes in total
na_only <- c() # Discard patient if all data are missing
wrong_order <- c() # Discard patient if episodes are not ordered correctly
no_common <- c() # Discard patient if any one episode lacks inter-episode data 
suitable <- c()

for(i in 1:length(y.dfs)) {
  y.df <- y.dfs[[i]]
  
  # no recurrence, skip
  if(length(unique(y.df$Episode_Identifier)) == 1) {
    no_recur <- c(no_recur, i)
    next
  }
  
  # >8 genotypes considered too complex
  if(nrow(y.df) > 8) {
    too_many_genos <- c(too_many_genos, i)
    next
  }
  
  # only use markers for which there is at least one non-NA
  ms <- MSs_all[apply(!is.na(y.df[MSs_all]), 2, any)]
  if(length(ms) == 0) {
    na_only <- c(na_only, i)
    next
  }
  
  # ensure that Episode is non-decreasing
  if(!all(order(y.df$Episode) == 1:nrow(y.df))) {
    wrong_order <- c(wrong_order, i)
    next
  }
  
  # Ensure for all episodes there are inter-episode data on at least one common marker
  y_by_episode <- plyr::dlply(y.df, 'Episode') # Group data by episode
  
  # Check there's at least one common marker with inter-episode data per episode 
  some_common_per_epi <- sapply(names(y_by_episode), function(epi) {
    
    other_epis <- setdiff(names(y_by_episode), epi) # Other episode names
    
    some_common <- any(sapply(MSs_all, function(ms){
      
      # Are the alleles at marker ms in epi all NAs? 
      per_marker_epi_data_all_na <- all(is.na(y_by_episode[[epi]][ms]))
      
      # Are the alleles at marker ms in the other episodes all NAs? 
      per_marker_other_epis_data_all_na <- all(sapply(other_epis, function(other_epi) {
        all(is.na(y_by_episode[[other_epi]][ms]))}))
      
      return(!per_marker_epi_data_all_na & !per_marker_other_epis_data_all_na) 
    })) 
    
    return(some_common)
  })
  
  if (!all(some_common_per_epi)) {
    no_common <- c(no_common, i)
    next
  }
    
  suitable <- c(suitable, i)
}
writeLines(paste("No recurrences:", length(no_recur)))
writeLines(paste("More than 8 genotypes (too complex):", length(too_many_genos)))
writeLines(paste("NAs only:", length(na_only)))
writeLines(paste("Episode numbers not in increasing order:", length(wrong_order)))
writeLines(paste("Not all episodes have inter-episode data:", length(no_common)))
writeLines(paste("Can proceed with inference:", length(suitable)))
```

Using a uniform prior over recurrence states, we perform inference on the recurrence states for the first 5 cases:

```{r}
results <- list()
for(i in suitable[1:5]) {
  y.df <- y.dfs[[i]]
  patient.id <- names(y.dfs)[i]
  writeLines(paste("ID:", patient.id))
  
  # only use markers for which there is at least one non-NA
  ms <- MSs_all[apply(!is.na(y.df[MSs_all]), 2, any)]
  
  # transform data frame format to format taken by compute_posterior()
  y.by.episode <- plyr:: dlply(y.df, 'Episode_Identifier')
  y <- lapply(y.by.episode, function(episode) {
    setNames(lapply(ms, function(m) {
      alleles <- episode[m][!is.na(episode[m])] # extract non-NAs
      if(length(alleles) > 0) return(alleles)
      return(NA) # if all are NAs, change empty vector to NA
    }), ms)
  })
  
  # Recurrent state genetic inference
  results[[patient.id]] <- compute_posterior(y, Fs_Combined)
}
```

