---
title: "Explore treatment of intra-episode siblings"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Explore treatment of intra-episode siblings}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  out.width = "100%", # Line width taken up by plot
  fig.dpi = 300, # For quality 
  fig.width = 7, # Increase to zoom out
  fig.height = 4, 
  cache = FALSE
)
```

```{r setup, echo = FALSE}
library(Pv3Rs)
# Access non-exported function
recombine_parent_ids <- utils::getFromNamespace("recombine_parent_ids", "Pv3Rs") 
```

As the following example illustrates, Pv3Rs sums over graphs with cliques of
three or more intra-episode siblings when MOIs used by the Pv3Rs model exceed
two. However, when the data input via the `y` argument of `compute_posterior()`
are as diverse as the MOIs used by the Pv3Rs model, graphs with cliques of three
or more intra-episode sibling parasites have zero likelihood; when MOIs used by
the Pv3Rs model are more diverse than the data input via the `y` argument of
`compute_posterior()` (e.g., because the user inputs elevated MOIs into
`compute_posterior()`), graphs with cliques of three or more intra-episode
siblings have non-zero likelihoods. In either case, provided MOIs used by the
Pv3Rs model are based on estimates from per-episode parasite genetic data
generated and analysed in bulk (i.e., not single-cell data), groups of three or
more not-half intra-episode siblings collapse to groups of two because

- MOI estimates based on maximum per-marker allele counts are at most two for
groups of three or more not-half siblings because groups of not-half siblings 
can draw from at most two parental alleles per marker.

- MOI estimates based on heteroallelic marker counts count at most two for 
groups of three or more not-half siblings because data from a group of not-half
siblings can be at most as diverse across markers as the two parental genotypes
from which they draw.

## Example of a group of four intra-episode siblings collapsing to two

Before showing how groups of intra-episode siblings collapse to pairs of
intra-episode siblings, we write a function to simulate data of a given allelic
richness (marker cardinality) on an enrolment episode comprising a stranger plus
a group of four siblings, two from one oocyst, two from another, all drawing
from the same two unrelated parental genotypes, and a recurrence with one
sibling. Technically, the enrolment episode contains five genetically distinct
genotypes and thus has a MOI of five.

```{r}
simulate_data <- function(marker_cardinality){
  
  # Magic numbers / quantities
  set.seed(5) # For reproducibility
  n_markers <- 200 # Number of markers
  n_strangers <- 3 # Number of stranger parasites
  n_oocysts <- 2 # Number of oocysts to draw from
  
  # Derived quantities
  alleles <- letters[1:marker_cardinality]
  markers <- paste0("m", 1:n_markers) # Marker names
  
  # Uniform allele frequencies 
  fs <- sapply(markers, simplify = FALSE, 
               function(m) setNames(rep(1/marker_cardinality, marker_cardinality), alleles))
  
  # Sample strangers
  strangers <- sapply(1:n_strangers, function(i) {
    sapply(markers, function(t) sample(names(fs[[t]]), size = 1, prob = fs[[t]]))
  })
  
  # Designate strangers
  parents <- strangers[, 1:2]
  
  # Map the markers to chromosomes. Assume equally sized chromosomes â€” reasonable
  # if and only if we later assume an equal number of crossovers per chromosome
  chrs_per_marker <- round(seq(0.51, 14.5, length.out = n_markers))
  
  # Sample parental allocations dependently per-oocyst 
  cs <- lapply(1:n_oocysts, function(o) recombine_parent_ids(chrs_per_marker))
  
  # Construct children from parental allocations 
  all_children <- lapply(1:n_oocysts, function(o) {
    oocyst_chidren <- sapply(1:n_markers, function(i) {
      sapply(1:ncol(cs[[o]]), function(j) parents[i,cs[[o]][i,j]])
    })
    colnames(oocyst_chidren) <- markers
    return(oocyst_chidren)
  })
  
  # Make enrolment infection
  enrol <- apply(rbind(all_children[[1]][1:2,], 
                       all_children[[2]][1:2,], 
                       strangers[,3]), 2, unique, simplify = F)
  
  # Make paired data
  data <- list(enrol = enrol, recur = as.list(all_children[[1]][1,]))
  
  return(list(data = data, fs = fs))
}
```
However, MOI estimates based on maximum per-marker allele counts are 
three and one when markers are polyallelic:

```{r}
polyallelic <- simulate_data(10) 
determine_MOIs(polyallelic$data)
```

And two and one when markers are biallelic: 

```{r}
biallelic <- simulate_data(2) 
determine_MOIs(biallelic$data)
```

Suppose we estimate MOIs of 3 and 1 for the biallelic data using an external
software that exploits heteroallelic marker counts (rather than the maximum
number of per-marker alleles), and input those external estimates into
`compute_posterior`. Providing data are simulated on a large number of markers
(200 above), we recover almost exactly the same posterior probabilities using
both the

- polyallelic data without user-specified MOIs
- biallelic data with user-specified MOIs three and one

```{r}
ppost <- suppressMessages(compute_posterior(y = polyallelic$data, 
                                            fs = polyallelic$fs, 
                                            return.logp = T))
bpost <- suppressMessages(compute_posterior(y = biallelic$data, 
                                            fs = biallelic$fs, 
                                            MOIs = c(3,1), return.logp = T))

ppost$marg
bpost$marg
```

However, given the polyallelic data the likelihoods of the following graphs with cliques of three intra-episode siblings are zero: 

```{r echo = FALSE, fig.height = 1.5} 
MOIs <- c(3,1) # MOIs based on data 
gs <- paste0("g", 1:sum(MOIs)) # Genotype names
ts_per_gs <- rep(1:length(MOIs), MOIs) # Episode indices per genotype
llikes <- sapply(ppost$RGs, function(RG) RG$logp) # Extract log likelihoods
par(mar = c(0, 0.5, 0, 0.5), mfrow = c(1,5)) # Plotting parameters
for(RG in ppost$RGs[which(is.infinite(llikes))]) {
plot_RG(RG_to_igraph(RG, MOIs),
        vertex_palette = "Greys",
        edge.curved = 0.08,
        vertex.size = 40,
        edge.width = 3,
        vertex.frame.width = 2,
        vertex.label.cex = 1,
        vertex.label.color = "black")
}
```

Whereas, all relationships have non-zero likelihoods given the biallelic data:

```{r} 
llikes <- sapply(bpost$RGs, function(RG) RG$logp) # Extract log likelihoods
any(is.infinite(llikes)) # Are there any minus infinity log likelihoods?
```

# Contribution to upper bounds 

The summation over graphs with cliques or three or more siblings is possibly
redundant given these graphs have no practical support. Even when these graphs
have zero likelihood, they contribute to maximum probabilities of recrudescence
/ reinfection through the conditionally uniform prior on graphs. However, for
upper bounds on the probability of reinfection / recrudescence given a single
recurrence the contribution is very small, especially for reinfection:

```{r echo = FALSE}
all_MOIs_chr <- colnames(maxima)
all_MOIs <- sapply(all_MOIs_chr, function(x) as.numeric(strsplit(x, split = "")[[1]]))
maxMOIs <- sapply(all_MOIs, function(x) max(x))
nMOIs <- sapply(all_MOIs, length) # Number of episodes
condition <- c(maxMOIs > 2 & nMOIs == 2) # MOIs > 2 (possibly of diff) and single recurrence
```

```{r echo = FALSE}
diffs <- maxima["I_with", condition] - maxima["I_wout", condition]
plot(NULL, xlim = c(0.5, 1), ylim = c(0.5,1), main = "Posterior bounds on reinfection", 
     xlab = "without summation", 
     ylab = "with summation"); abline(a = 0, b = 1)
points(y = maxima["I_with", condition], x = maxima["I_wout", condition], 
       pch = 21, bg = "red")
title(sub = sprintf("Largest difference: %s", round(max(abs(diffs)), 4))) 
```


```{r echo = FALSE}
diffs <- maxima["C_with", condition] - maxima["C_wout", condition]
max_diff <- max(abs(diffs))
cutoff <- 0.0175
graphs_with_big_diff <- names(all_MOIs)[condition][abs(diffs) > cutoff]
```

For recrudescence, `r sum(abs(diffs) > cutoff)` MOI vectors (circled
below) result in an absolute difference greater than `r cutoff`. They are those
that feature large MOIs (MOI vectors `r graphs_with_big_diff`) and thus those
whose graph spaces include graphs with the largest cliques of intra-episode
siblings.

```{r echo = FALSE}
# Compare with and without summation
plot(NULL, xlim = c(0.5, 1), ylim = c(0.5,1), main = "Posterior bounds on recrudescence", 
     xlab = "without summation", ylab = "with summation"); abline(a = 0, b = 1)
points(y = maxima["C_with", condition][abs(diffs) > cutoff], 
       x = maxima["C_wout", condition][abs(diffs) > cutoff], pch = 1, cex = 2)
points(y = maxima["C_with", condition], x = maxima["C_wout", condition], 
       pch = 21, bg = "yellow")

title(sub = sprintf("Largest difference: %s", round(max_diff, 4))) 

text(y = maxima["C_with", condition][abs(diffs) > cutoff], 
     x = maxima["C_wout", condition][abs(diffs) > cutoff], 
     labels = names(all_MOIs)[condition][abs(diffs) > cutoff], 
     pos = c(3,1,1,3,1), cex = 0.5, offset = 1)
```


The contribution is likely smaller for all vectors of three or more MOIs that
are computationally feasible under Pv3Rs because their constituent MOIs are
necessarily less than 7 (computationally feasible MOI vectors are those where
the total genotype count is at most eight).
