<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Demonstrate Pv3Rs usage â€¢ Pv3Rs</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Demonstrate Pv3Rs usage">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">Pv3Rs</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/demonstrate-usage.html">Demonstrate Pv3Rs usage</a></li>
    <li><a class="dropdown-item" href="../articles/enumerate.html">Enumerate graphs and partitions</a></li>
    <li><a class="dropdown-item" href="../articles/intra-episode-siblings.html">Explore treatment of intra-episode siblings</a></li>
    <li><a class="dropdown-item" href="../articles/understand-graph-prior.html">Understand graph prior ramifications</a></li>
    <li><a class="dropdown-item" href="../articles/understand-half-sibs.html">Understand half-sibling misspecification</a></li>
    <li><a class="dropdown-item" href="../articles/understand-posterior.html">Understand posterior estimates</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/aimeertaylor/Pv3Rs/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Demonstrate Pv3Rs usage</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/aimeertaylor/Pv3Rs/blob/main/vignettes/demonstrate-usage.Rmd" class="external-link"><code>vignettes/demonstrate-usage.Rmd</code></a></small>
      <div class="d-none name"><code>demonstrate-usage.Rmd</code></div>
    </div>

    
    
<p>In this vignette we demonstrate the basic <code>Pv3Rs</code> work
flow for a single study participant; i.e.,</p>
<ol style="list-style-type: decimal">
<li>Plot data</li>
<li>Estimate recurrent state probabilities</li>
<li>Plot probability estimates</li>
</ol>
<p>And we show how one might explore relationship graphs and their log
likelihoods. In one or more upcoming scientific publications, we plan to
publish a more elaborate study-level <code>Pv3Rs</code> tutorial
including examples of</p>
<ul>
<li>how to generate estimates pairwise when the total genotype count
over multiple recurrences exceeds eight</li>
<li>how to compute false discovery rates</li>
<li>how to do a sensitivity analysis for genotyping errors</li>
<li>how to do a sensitivity analysis for sibling misspecification</li>
<li>an alternative genetic distance based approach using whole-genome
sequence data</li>
<li>how posterior probabilities (rather than binary classifications) can
be used estimate treatment efficacy</li>
<li>an estimation of a false negative rate</li>
</ul>
<div class="section level2">
<h2 id="basic-workflow">Basic workflow<a class="anchor" aria-label="anchor" href="#basic-workflow"></a>
</h2>
<p>We first look at a synthetic example of three episodes (episode names
are optional) with three markers (marker names are obligatory) whose
alleles have known frequencies.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Enrollment"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'C'</span>,<span class="st">'G'</span>,<span class="st">'T'</span><span class="op">)</span>, </span>
<span>                              m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'C'</span><span class="op">)</span>,</span>
<span>                              m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'C'</span>,<span class="st">'G'</span>,<span class="st">'T'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          <span class="st">"Recurrence 1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'C'</span>,<span class="st">'T'</span><span class="op">)</span>, </span>
<span>                                m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span><span class="op">)</span>, </span>
<span>                                m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>,<span class="st">'C'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>          <span class="st">"Recurrence 2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'T'</span><span class="op">)</span>, </span>
<span>                                m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span><span class="op">)</span>, </span>
<span>                                m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>m1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.27</span>, C <span class="op">=</span> <span class="fl">0.35</span>, G <span class="op">=</span> <span class="fl">0.18</span>, T <span class="op">=</span> <span class="fl">0.20</span><span class="op">)</span>,</span>
<span>           m2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.78</span>, C <span class="op">=</span> <span class="fl">0.14</span>, G <span class="op">=</span> <span class="fl">0.07</span>, T <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>           m3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>A <span class="op">=</span> <span class="fl">0.21</span>, C <span class="op">=</span> <span class="fl">0.45</span>, G <span class="op">=</span> <span class="fl">0.26</span>, T <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We plot the data using <code><a href="../reference/plot_data.html">plot_data()</a></code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_data.html">plot_data</a></span><span class="op">(</span>ys <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"Participant data"</span> <span class="op">=</span> <span class="va">y</span><span class="op">)</span>, fs <span class="op">=</span> <span class="va">fs</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Different colours in the plotted data (left) represent different
alleles. The legend (right) shows all alleles per marker with one column
per marker. The order of the markers in the legend is the same as that
in the plotted data (see horizontal axis, left). When allele frequencies
are specified, the heights of differently coloured areas in the legend
are proportional to allele frequencies, s.t. rare alleles have
relatively small legend areas (e.g., T at m2 is rare).</p>
<p><strong>Aside</strong> Given the maximum number of alleles observed
per-episode, the most parsimonious explanation is that there are 3, 2, 1
distinct parasite genotypes in the three episodes respectively (a
genotype is here defined as a realisation of the haploid parasite genome
representing a group of clonal parasites). In this synthetic example,
markers are quart-allelic for brevity; this imposes a rather low upper
bound on MOI estimates based on maximum per-marker allele counts. In
reality, more diverse markers are recommended for recurrent state
inference. If better MOI estimates were available based on more diverse
markers, the data from those more diverse markers could be added to the
data input <code>y</code> of <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code>. If better
MOI estimates based on heteroallelic marker counts across many markers
were available, they could be used in recurrent state inference by
specifying the <code>MOI</code> argument in the function
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code>.</p>
<p>When performing genetic recurrent state inference, the bulk of the
computational time lies in computing log-likelihoods of graphs of
relationships between genotypes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">fs</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of valid relationship graphs (RGs) is 1315</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Computing log p(Y|RG) for 1315 RGs</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Finding log-likelihood of each vector of recurrent states</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span></code></pre></div>
<p>In the call to <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> above we did not
specify a prior, and so a uniform prior across all three recurrence
states per recurrence is assumed by default. Marginal posterior
probabilities (probabilities of recrudescence <code>C</code>, relapse
<code>L</code>, and reinfection <code>I</code> for each recurrence) are
stored in <code>post$marg</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span><span class="op">$</span><span class="va">marg</span></span>
<span><span class="co">#&gt;                     C         L          I</span></span>
<span><span class="co">#&gt; Recurrence 1 0.000000 0.3294555 0.67054446</span></span>
<span><span class="co">#&gt; Recurrence 2 0.670205 0.2388744 0.09092065</span></span></code></pre></div>
<p>Marginal posterior probabilities can be plotted on the simplex using
the function <code><a href="../reference/plot_simplex.html">plot_simplex()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot simplex</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_simplex.html">plot_simplex</a></span><span class="op">(</span>v_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>C <span class="op">=</span> <span class="st">"Recrudescence"</span>, </span>
<span>                          L <span class="op">=</span> <span class="st">"Relapse"</span>, </span>
<span>                          I <span class="op">=</span> <span class="st">"Reinfection"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">marg</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Project marginal posterior probabilties onto 2D coordinates</span></span>
<span><span class="va">xy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">marg</span>, <span class="fl">1</span>, <span class="va">project2D</span><span class="op">)</span> </span>
<span><span class="va">xy_prior</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/project2D.html">project2D</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Plot projection on the simplex</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xy</span><span class="op">[</span><span class="st">"x"</span>,<span class="op">]</span>, y <span class="op">=</span> <span class="va">xy</span><span class="op">[</span><span class="st">"y"</span>,<span class="op">]</span>, pch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xy</span><span class="op">[</span><span class="st">"x"</span>,<span class="op">]</span>, y <span class="op">=</span> <span class="va">xy</span><span class="op">[</span><span class="st">"y"</span>,<span class="op">]</span>, pos <span class="op">=</span> <span class="fl">3</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">marg</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add default prior per recurrence</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/points.html" class="external-link">points</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xy_prior</span><span class="op">[</span><span class="st">"x"</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">xy_prior</span><span class="op">[</span><span class="st">"y"</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xy_prior</span><span class="op">[</span><span class="st">"x"</span><span class="op">]</span>, y <span class="op">=</span> <span class="va">xy_prior</span><span class="op">[</span><span class="st">"y"</span><span class="op">]</span>, pos <span class="op">=</span> <span class="fl">3</span>, labels <span class="op">=</span> <span class="st">"Prior probability"</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The point in the yellow region is most likely a recrudescence with
posterior probability greater than 0.5 (it falls in the bright yellow
region); the point in the red region is most likely a reinfection with
posterior probability greater than 0.5 (it falls in the bright red
region).</p>
<p>Joint posterior probabilities (probabilities of chronological
sequences of recrudescence <code>C</code>, relapse <code>L</code>, and
reinfection <code>I</code>) are stored in <code>post$joint</code>. Here,
we find the most likely sequence of recurrence states is
<code>r names(which.max(post$joint))</code> with posterior probability
<code>r post$joint[which.max(post$joint)]</code></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span><span class="op">$</span><span class="va">joint</span></span>
<span><span class="co">#&gt;         CC         LC         IC         CL         LL         IL         CI </span></span>
<span><span class="co">#&gt; 0.00000000 0.21338194 0.45682305 0.00000000 0.08501504 0.15385932 0.00000000 </span></span>
<span><span class="co">#&gt;         LI         II </span></span>
<span><span class="co">#&gt; 0.03105856 0.05986209</span></span></code></pre></div>
<p>Aside: we do not recommend running <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code>
for data whose total genotype count (sum of per-episode multiplicities
of infection) exceeds eight. That said, we have not imposed a hard limit
on the code. It is possible, but very long, to generate estimates using
data with a total genotype count up to 10 (see code below, which we time
out after 2 seconds); above 10 genotypes, calls to
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> are liable cause memory-use problems
and fail.</p>
</div>
<div class="section level2">
<h2 id="exploration-of-relationship-graphs-and-their-log-likelihoods">Exploration of relationship graphs and their log likelihoods<a class="anchor" aria-label="anchor" href="#exploration-of-relationship-graphs-and-their-log-likelihoods"></a>
</h2>
<p>If we want to explore relationship graphs and their log likelihoods
we need to set both <code>return.RG</code> and <code>return.logp</code>
to <code>TRUE</code> when computing the posterior; they are
<code>FALSE</code> by default (see below).</p>
<p>To compute the posterior, summations over per-marker allelic
assignments that are equivalent up to within-episode genotype
permutations are redundant. As such, by default,
<code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> does not sum over them, conserving both
memory and compute time; see <code><a href="../reference/enumerate_alleles.html">enumerate_alleles()</a></code>. The
exploitation of permutation symmetry requires a scheme to choose a
single representative among permutations that are otherwise equivalent.
When user-specified MOIs are greater than those based on per-marker
allele counts, <code><a href="../reference/compute_posterior.html">compute_posterior()</a></code> sums over all
permutations because the scheme is too complicated. Likewise, to compute
meaningful graph likelihood values (i.e., values that are not dependent
on the aforementioned representative-choosing scheme), all permutations
are summed over when <code>return.logp = TRUE</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_posterior.html">compute_posterior</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">fs</span>, return.RG <span class="op">=</span> <span class="cn">TRUE</span>, return.logp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Number of valid relationship graphs (RGs) is 1315</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Computing log p(Y|RG) for 1315 RGs</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span>
<span><span class="co">#&gt; Finding log-likelihood of each vector of recurrent states</span></span>
<span><span class="co">#&gt; =============================================================================|</span></span></code></pre></div>
<p>We recover the same posterior as before</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">post</span><span class="op">$</span><span class="va">joint</span></span>
<span><span class="co">#&gt;         CC         LC         IC         CL         LL         IL         CI </span></span>
<span><span class="co">#&gt; 0.00000000 0.21338194 0.45682305 0.00000000 0.08501504 0.15385932 0.00000000 </span></span>
<span><span class="co">#&gt;         LI         II </span></span>
<span><span class="co">#&gt; 0.03105856 0.05986209</span></span></code></pre></div>
<p>The log-likelihood of each relationship graph is also returned in the
output. We plot the relationship graph(s) with the largest
likelihood.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract all log likelihoods</span></span>
<span><span class="va">lliks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">post</span><span class="op">$</span><span class="va">RGs</span>, <span class="kw">function</span><span class="op">(</span><span class="va">RG</span><span class="op">)</span> <span class="va">RG</span><span class="op">$</span><span class="va">logp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract the relationship graphs (RGs) with the largest log likelihood</span></span>
<span><span class="va">RGs</span> <span class="op">&lt;-</span> <span class="va">post</span><span class="op">$</span><span class="va">RGs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">lliks</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">lliks</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Generate genotype names </span></span>
<span><span class="va">gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"g"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Generate episode indices for genotypes</span></span>
<span><span class="va">ts_per_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, <span class="fu"><a href="../reference/determine_MOIs.html">determine_MOIs</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># The sequences of states compatible with RGs</span></span>
<span><span class="va">seqs_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">RGs</span>, <span class="va">compatible_rstrs</span>, gs_per_ts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">gs</span>, <span class="va">ts_per_gs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot RG</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">4</span><span class="op">)</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">RGs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/plot_RG.html">plot_RG</a></span><span class="op">(</span><span class="fu"><a href="../reference/RG_to_igraph.html">RG_to_igraph</a></span><span class="op">(</span><span class="va">RGs</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="va">gs</span>, <span class="va">ts_per_gs</span><span class="op">)</span>, vertex.size <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/box.html" class="external-link">box</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># Generate infection colours in order to add a legend</span></span>
<span><span class="va">infection_colours</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">8</span>, <span class="st">"Set2"</span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Add a legend</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"bottomright"</span>, pch <span class="op">=</span> <span class="fl">21</span>, pt.bg <span class="op">=</span> <span class="va">infection_colours</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ts_per_gs</span><span class="op">)</span><span class="op">]</span>, </span>
<span>       bty <span class="op">=</span> <span class="st">"n"</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, title <span class="op">=</span> <span class="st">"Episode"</span><span class="op">)</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-10-1.png" width="672" style="display: block; margin: auto;"></p>
<p>N.B. There are two symmetric graphs with maximum likelihood: they are
isomorphic up to within-episode genotype permutations. The user should
note that, as in the current example, the relationship graph(s) with the
largest likelihood (e.g., those above) might not be in the equivalence
class for which the data are most probable when all relationship graphs
in that class are summed over (e.g., those below).</p>
<p>The user should also note that,</p>
<ul>
<li><p>the maximum-likelihood graph(s) might be incompatible with the
maximum-posterior state sequence (albeit not true of the current
example: the graphs above are compatible with IC)
<!-- see example in DevFiles/Investigate_Symmetries --></p></li>
<li><p>the graphs in the maximum-likelihood equivalence class might be
incompatible with the maximum-posterior state sequence (true of the
current example, given the class with the largest likelihood contains
graphs with sibling edges that are incompatible with IC).</p></li>
</ul>
<!-- We suspect one of these cases above has to be false, but we cannot rule out
the possibility that both are true without further exploration --><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sorted_lliks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">lliks</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co"># Sort log likelihoods</span></span>
<span><span class="va">adj_equal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">sorted_lliks</span>, lag <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="va">.Machine</span><span class="op">$</span><span class="va">double.eps</span><span class="op">^</span><span class="fl">0.5</span> <span class="co"># Find matches</span></span>
<span><span class="va">decr_idxs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">adj_equal</span> <span class="op">==</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Change points: 2, 8, 14, 20, 32, ...</span></span>
<span><span class="va">class_sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">decr_idxs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/diff.html" class="external-link">diff</a></span><span class="op">(</span><span class="va">decr_idxs</span><span class="op">)</span><span class="op">)</span> <span class="co"># Number of graphs per class</span></span>
<span></span>
<span><span class="co"># log likelihood of representative from each 'equivalence class' (EC)</span></span>
<span><span class="va">lliks_unique</span> <span class="op">&lt;-</span> <span class="va">sorted_lliks</span><span class="op">[</span><span class="va">decr_idxs</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># EC likelihood</span></span>
<span><span class="va">class_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">lliks_unique</span><span class="op">)</span><span class="op">*</span><span class="va">class_sizes</span></span>
<span><span class="va">max_class_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">class_ps</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">class_ps</span><span class="op">)</span><span class="op">)</span> <span class="co"># ML EC index </span></span>
<span><span class="va">max_idx</span> <span class="op">&lt;-</span> <span class="va">decr_idxs</span><span class="op">[</span><span class="va">max_class_p</span><span class="op">]</span> <span class="co"># Index of last graph in ML EC</span></span>
<span><span class="va">max_size</span> <span class="op">&lt;-</span> <span class="va">class_sizes</span><span class="op">[</span><span class="va">max_class_p</span><span class="op">]</span> <span class="co"># Number of graphs in ML EC</span></span>
<span></span>
<span><span class="co"># Plot all graphs within the ML EC </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0.1</span>,<span class="fl">4</span><span class="op">)</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">RG_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">lliks</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="co"># order RGs by logl</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="op">(</span><span class="va">max_idx</span><span class="op">-</span><span class="va">max_size</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="va">max_idx</span><span class="op">)</span> <span class="op">{</span> <span class="co"># EC consists of the RGs with logl rank 21-32</span></span>
<span>  <span class="va">RG</span> <span class="op">&lt;-</span> <span class="va">post</span><span class="op">$</span><span class="va">RGs</span><span class="op">[[</span><span class="va">RG_order</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="va">RG_igraph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RG_to_igraph.html">RG_to_igraph</a></span><span class="op">(</span><span class="va">RG</span>, <span class="va">gs</span>, <span class="va">ts_per_gs</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/plot_RG.html">plot_RG</a></span><span class="op">(</span><span class="va">RG_igraph</span>, vertex.size <span class="op">=</span> <span class="fl">25</span>, vertex.label <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/box.html" class="external-link">box</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="demonstrate-usage_files/figure-html/unnamed-chunk-11-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aimee Taylor, Yong See Foo, European Union, Project 101110393 â€” PvRecur.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
